### ISSUE TITLE
TADA TTS: talker decodes greedily with no repetition penalty (diverges from the HumeAI reference)

### ISSUE THREAD

**Issue by thiswillbeyourgithub** [CrispStrobe/CrispASR#197]
Human: Hi, congrants and many thanks for this new release. I gave it a go as I was intrigued by the performance of the TADA model. I gotta say it's pretty promising as the text generation is really fast and the voice cloning is really close to the original voice IMO. But I kept running into hallucinations issues (extra noises added, some mini-screams here and there, etc) so I asked claude code to look for knobs to turn and apparently there are some little pipes that failed to get connected along the way. Hence I asked claude to produce this dry inhuman text below to humbly request for this new plumbing. Thank you very much for CrispASR, it's really terrific.

---

## What I need
Make the TADA talker decode like the HumeAI reference: repetition penalty + temperature + top-k/top-p, instead of pure greedy argmax. Plus expose the flow-matching knobs that are currently frozen at init. Defaults should equal upstream `InferenceOptions`.

## Why
With `tada-3b-ml`, `-l fr`, and a cloned French voice, output regularly hallucinates: wrong/repeated words, looping, mangled liaisons and punctuation. `TADA_NUM_CANDIDATES` does not help (even at 1000), because it only tunes the duration flow-matching head, not the content. The real cause is in the talker decoder, and right now none of it is tunable at runtime.

## The mistakes (current `main`)
1. The talker decodes **purely greedily**: `src/tada_tts.cpp:2625` is an unconditional `argmax_logits(...)` (its own comment says "Next token prediction (greedy)."). Greedy AR TTS with no repetition control loops and hallucinates, worst on lower-confidence non-English text.
2. **No repetition penalty, no top-p, no top-k** anywhere in the generation loop.
3. The `temperature` knob is **dead code**: the field is declared (`src/tada_tts.h:30`), defaulted to `0.0` (`src/tada_tts.cpp:1727`), and has a public setter `tada_set_temperature()` (`src/tada_tts.cpp:1955`), but it is never read by the token-selection step. So the one knob that looks available is a no-op.
4. This diverges from the reference. `tada/modules/tada.py` `InferenceOptions` defaults: `text_do_sample=True, text_temperature=0.6, text_top_k=0, text_top_p=0.9, text_repetition_penalty=1.1`. CrispASR effectively runs `do_sample=False` with no penalty. The acoustic/FM defaults (`acoustic_cfg=1.6`, `noise_temp=0.9`, `num_acoustic_candidates`) DO match, so the gap is specifically the text decoder.
5. Several model knobs are frozen at init with no setter/env (`src/tada_tts.h:32-37`): `acoustic_cfg`, `noise_temp`, `num_fm_steps`, `max_tokens`. And `duration_cfg` (1.0) and `cfg_schedule` (cosine) are hardcoded literals in `src/tada_tts.cpp` (not even parameters). Cannot experiment without recompiling.

## Proposed fix
- **Primary:** at the token-selection step (`src/tada_tts.cpp:2625`), apply repetition penalty over the already-generated ids, then temperature + top-k/top-p, then sample; fall back to argmax when `do_sample` is false or `temperature == 0`. Wire the existing `temperature` field. Add `text_top_p`, `text_top_k`, `text_repetition_penalty`, `text_do_sample` to `tada_context_params` with the upstream defaults above.
- **Plumbing:** expose via env vars + C ABI setters + server `/v1/audio/speech` params, mirroring the existing `TADA_NUM_CANDIDATES` / `tada_set_num_candidates()` pattern (`src/tada_tts.h:64`). Suggested env: `TADA_TEMPERATURE`, `TADA_TOP_P`, `TADA_TOP_K`, `TADA_REPETITION_PENALTY`, `TADA_DO_SAMPLE`.
- **Secondary (easy, same PR or follow-up):** add setters/env for the frozen knobs `acoustic_cfg`, `noise_temp`, `num_fm_steps`, `max_tokens`, and make `duration_cfg` / `cfg_schedule` parameters.

## Acceptance check
Same French text + cloned ref, new defaults vs current `main`: looping/hallucination gone; `--seed` now changes wording (proves sampling is live); raising `TADA_REPETITION_PENALTY` measurably reduces repeats.

---
Investigated and written with help from Claude Code.

**Comment by CrispStrobe** [comment]
thanks! could you build from latest main and retest?

**Comment by thiswillbeyourgithub** [comment]
Thanks! I'm building it right now. If I'm reading your commits correctly you improved the sampler but it would be best if we could change the values at query time instead of reloading the entire docker container for that.

**Comment by CrispStrobe** [comment]
Good call — it's wired now (on `main`). The sampler is per-request on `POST /v1/audio/speech`, no restart needed: `temperature`, `top_p`, `top_k`, `repetition_penalty`, `do_sample`, and `num_candidates` are all accepted as JSON fields. Any field you omit falls back to whatever the server was started with (env / flags), and requests don't leak settings into each other.

```bash
curl -s http://localhost:8080/v1/audio/speech \
  -H 'Content-Type: application/json' \
  -d '{"input":"one two three four","temperature":0.6,"top_p":0.9,"top_k":0,"repetition_penalty":1.1,"do_sample":true,"num_candidates":4}' \
  -o out.wav
```

The same knobs are also exposed as session setters (`set_top_k` / `set_do_sample` / `set_top_p` / `set_repetition_penalty` / `set_temperature` / `set_tts_num_candidates`) across all the language bindings if you're embedding rather than going through the HTTP server.

**Comment by thiswillbeyourgithub** [comment]
Great. Seems to work fine. The defaults sound indeed better I think but can't prove it.

~~Ultimately it's gonna be necessary to add a mechanism to change the voice at query time though. Is that doable?~~ Moved to its own issue: #201

**Comment by thiswillbeyourgithub** [comment]
On second look, it seems like there are some inference parameters left to implement:

https://github.com/HumeAI/tada/blob/d6b72c3329691ea087693ca175e203a4b1a3aebf/tada/modules/tada.py#L23


According to [this issue](https://github.com/HumeAI/tada/issues/26) it seems to influence the output quite a lot.

FWIW I'm especially interested in slow accurate generations and this model seems to have the right knobs to choose between "quick and dirty" and "slow and accurate"

**Comment by thiswillbeyourgithub** [comment]
After more testing, I decided to do some tests with a python script from tada. I don't know if it's because of the fact that the knobs are exposed and I can tweak them or if there's something legitimately wrong with your code but on my tests the outputs from CrispASR consistently add long pauses and hums and silences when there's punctuation, the py script does not. For example if I use `Hi. My name is Bob. It's really nice to meet you!` I seem to get long pauses and wind like noises when there are periods.

This might totally be me cherrypicking my own senses of course but I preferred to notify you just in case.

**Comment by CrispStrobe** [comment]
Thanks. I see that TADA is atm still a work in progress. Fixes already attempted:

**1. The punctuation pauses / hums** The CLI and server were splitting your text at every period and synthesizing each sentence in isolation, then concatenating with inserted silence. That diverges from `tada.py`, whose AR talker generates the whole utterance in a single pass. The damage was worst on short sentences: `"Hi."` alone made the per-token duration head over-predict the trailing pause — ~500 frames (~9 s) of silence + that wind/hum — and then more silence got padded between chunks. On your exact example:

```
split (old):  12.70 s, duration spike 251 frames, ASR even drops the final "you!"
single-pass:   5.22 s, natural max ~57 frames,    ASR verbatim incl. "you"
```

TADA now feeds the whole text in one shot like the reference (same policy we already use for VibeVoice/Qwen3-TTS), so the punctuation pauses and hums are gone.

**2. The remaining `InferenceOptions` knobs are now query-time tunable.** `num_flow_matching_steps`, `acoustic_cfg` and `noise_temp` were init-only; they're now per-request on `POST /v1/audio/speech`, no restart:

```bash
# "slow and accurate": more ODE steps = crisper acoustics (the main lever)
curl -s http://localhost:8080/v1/audio/speech \
  -H 'Content-Type: application/json' \
  -d '{"input":"Hi. My name is Bob. It'\''s really nice to meet you!","num_steps":25,"cfg_scale":1.6,"noise_temp":0.9}' \
  -o out.wav
```

- `num_steps` → `num_flow_matching_steps` (default 10) — your "quick and dirty" vs "slow and accurate" axis. Measured on the 1B/Q4_K: 4 steps ≈ 77 s, 25 steps ≈ 185 s, both ASR-verbatim, the latter audibly crisper.
- `cfg_scale` → `acoustic_cfg` (default 1.6)
- `noise_temp` → `noise_temp` (default 0.9)

Same as the sampler knobs: omit a field and it falls back to the server's startup value (env/flags), and requests don't leak settings into each other. The env vars `TADA_NUM_FM_STEPS` / `TADA_ACOUSTIC_CFG` / `TADA_NOISE_TEMP` and the session setters (`set_tts_steps` → FM steps, `set_cfg_weight` → acoustic CFG, `set_tts_noise_temp`) are wired across all bindings too.

Please rebuild from `main` and let me know how the slow/accurate mode sounds for your use case.

More in the works and incoming...

**Comment by CrispStrobe** [comment]
Fixed — per-request talker sampler with top_k/do_sample/num_candidates wired through all consumers (0f3e6d2c). Single-pass whole-text generation stops punctuation pauses/hums (78ab0176). Acoustic flow-matching knobs exposed at query time (d4802770).
